File: pdu/python/sensor_msgs/pdu_cdr_conv_PointCloud2.py
from .pdu_pytype_PointCloud2 import PointCloud2
from ..pdu_cdr_runtime import CdrReader, CdrWriter

from ..std_msgs.pdu_cdr_conv_Header import *
from ..sensor_msgs.pdu_cdr_conv_PointField import *
from ..builtin_interfaces.pdu_cdr_conv_Time import *



def py_to_cdr_body_PointCloud2(writer: CdrWriter, src: PointCloud2):
    py_to_cdr_body_Header(writer, src.header)
    writer.write_uint32(src.height)
    writer.write_uint32(src.width)
    writer.write_sequence_length(src.fields)
    for elem in src.fields:
        py_to_cdr_body_PointField(writer, elem)
    writer.write_bool(src.is_bigendian)
    writer.write_uint32(src.point_step)
    writer.write_uint32(src.row_step)
    writer.write_sequence_length(src.data)
    for elem in src.data:
        writer.write_uint8(elem)
    writer.write_bool(src.is_dense)


def cdr_body_to_py_PointCloud2(reader: CdrReader, dst: PointCloud2):
    cdr_body_to_py_Header(reader, dst.header)
    dst.height = reader.read_uint32()
    dst.width = reader.read_uint32()
    dst.fields = []
    for _ in range(reader.read_uint32()):
        elem = PointField()
        cdr_body_to_py_PointField(reader, elem)
        dst.fields.append(elem)
    dst.is_bigendian = reader.read_bool()
    dst.point_step = reader.read_uint32()
    dst.row_step = reader.read_uint32()
    dst.data = [reader.read_uint8() for _ in range(reader.read_uint32())]
    dst.is_dense = reader.read_bool()
    return dst


def py_to_cdr_PointCloud2(src: PointCloud2) -> bytes:
    writer = CdrWriter()
    writer.write_encapsulation()
    py_to_cdr_body_PointCloud2(writer, src)
    return writer.bytes()


def cdr_to_py_PointCloud2(cdr_payload) -> PointCloud2:
    reader = CdrReader(cdr_payload)
    reader.read_encapsulation()
    dst = PointCloud2()
    return cdr_body_to_py_PointCloud2(reader, dst)

File: pdu/javascript/sensor_msgs/pdu_cdr_conv_PointCloud2.js
import { PointCloud2 } from './pdu_jstype_PointCloud2.js';
import { PduCdrWriter, PduCdrReader } from '../pdu_cdr_runtime.js';
import { Header } from '../std_msgs/pdu_jstype_Header.js';
import { PduHeaderConverter } from '../std_msgs/pdu_cdr_conv_Header.js';
import { PointField } from '../sensor_msgs/pdu_jstype_PointField.js';
import { PduPointFieldConverter } from '../sensor_msgs/pdu_cdr_conv_PointField.js';
import { Time } from '../builtin_interfaces/pdu_jstype_Time.js';
import { PduTimeConverter } from '../builtin_interfaces/pdu_cdr_conv_Time.js';


export class PduPointCloud2Converter {
    /**
     * @param {PduCdrWriter} writer
     * @param { PointCloud2 } src
     */
    static to_cdr_body(writer, src) {
        PduHeaderConverter.to_cdr_body(writer, src.header);
        writer.write_uint32(src.height);
        writer.write_uint32(src.width);
        writer.write_sequence_length(src.fields);
        for (const elem of src.fields) {
            PduPointFieldConverter.to_cdr_body(writer, elem);
        }
        writer.write_bool(src.is_bigendian);
        writer.write_uint32(src.point_step);
        writer.write_uint32(src.row_step);
        writer.write_sequence_length(src.data);
        for (const elem of src.data) {
            writer.write_uint8(elem);
        }
        writer.write_bool(src.is_dense);
    }

    /**
     * @param {PduCdrReader} reader
     * @param { PointCloud2 } dst
     * @returns { PointCloud2 }
     */
    static cdr_body_to_js(reader, dst) {
        PduHeaderConverter.cdr_body_to_js(reader, dst.header);
        dst.height = reader.read_uint32();
        dst.width = reader.read_uint32();
        dst.fields = [];
        for (let i = 0, len = reader.read_uint32(); i < len; i++) {
            dst.fields.push(PduPointFieldConverter.cdr_body_to_js(reader, new PointField()));
        }
        dst.is_bigendian = reader.read_bool();
        dst.point_step = reader.read_uint32();
        dst.row_step = reader.read_uint32();
        dst.data = Array.from({ length: reader.read_uint32() }, () => reader.read_uint8());
        dst.is_dense = reader.read_bool();
        return dst;
    }

    /**
     * @param { PointCloud2 } src
     * @returns {ArrayBuffer}
     */
    static to_cdr(src) {
        const writer = new PduCdrWriter();
        writer.write_encapsulation();
        this.to_cdr_body(writer, src);
        return writer.get_buf();
    }

    /**
     * @param {ArrayBuffer|ArrayBufferView} cdrPayload
     * @returns { PointCloud2 }
     */
    static from_cdr(cdrPayload) {
        const reader = new PduCdrReader(cdrPayload);
        reader.read_encapsulation();
        return this.cdr_body_to_js(reader, new PointCloud2());
    }
}
